### PR TITLE
Y devtools

### PR DESCRIPTION
**Pull request by jrowny**
- Adds Yjs to dev tools
- React-Flow document viewer
- StorageTree used for JSON view (this could use further optimization)
- ChangeLog viewer
- Awareness viewer

NOTE: requires change to Yjs-provider package so you will need to link locally to test

### CODE REVIEW

**Review by nvie — COMMENTED** [review]
This is starting to come together nicely! I mostly had a play with how it behaves, and haven't focused my review on the actual code too much just yet.

Below are some bugs/feedback/questions/suggestions.

Not all of them have to be addressed in this first release, but I think it would be nice to fine-tune these for a better actual debugging experience over time, for real world apps.

---

### Flickering bug?
Just being alone in the room, idle, not doing anything with the app, and just scrolling the Awareness tab, I noticed some "highlighted" rows that don't update. This means those rows are being unmounted and remounted unnecessarily?

https://github.com/liveblocks/liveblocks/assets/83844/b5a9fec2-b274-40af-8b92-de6539deb50a

---

### Infinite updates bug?
Being alone in a room, DevTools open, being idle, DevTools are idle too. Then, when you join another client and just take the tiniest action, it triggers a seemingly infinite loop of "me" and "others" updates, causing lots of flickering in the DevTools. Not sure if this is actually a bug in DevTools, or in the application, honestly.

https://github.com/liveblocks/liveblocks/assets/83844/566f4767-0d50-4dc8-a01f-94f044321a6d

---

### Document view
Maybe for later, but I found this view only somewhat useful in the case of large bodies of text. Is the key `default` here a key that users can control? I wondered where it was coming from, because I could find the key `"default"` in the source code of `nextjs-yjs-tiptap`. When working with text editors, text bodies are often large. Should we maybe not use the Arborist tree here, or at least make those text content rows multi-line, and wrapping?

Screen Shot 2023-09-20 at 10 37 09@2x

---

### Changes view (as list)
Could we visualize the `[object Object]` here? In general, I find these views rather hard to work with. I don't have too much hand-on experience with Yjs myself yet, but do these terms "DeleteSet", "ContentFormat", "ContentDeleted", "rX", etc make (enough) intuitive sense to people that are familiar with Yjs?

Screen Shot 2023-09-20 at 10 43 41@2x

What do these groupings mean? Are these the batches that might eventually get combined and regrouped as the result of garbage collection? I find this information a bit rough to make sense of still. I think we could offer tools that might make exploring this data easier. Maybe add toggles to hide "Garbage collection" or "ContentDeleted" nodes for a more condensed view?

---

### Changes view (as diagram)
Same feedback goes for the diagram view. I found the diagrams to look fancy, but I do not see myself using it much because it quickly becomes so large in practice that it's unwieldy and a bit cryptic. One thing that might help there is focusing on new nodes as they trickle in. For example, if I add text in the application, I have no idea where those new nodes are appearing in the diagram view. It would be nice if the newest nodes are scrolled into view, maybe, so you get a more direct debugging interaction. (Maybe make it possible to opt out of that behavior in case you want to focus on existing nodes, but it would be nice if it were the default.)

---

### Awareness view
I noticed that there was quite some overlap in the data inside Presence and inside the Awareness tabs (but not 100% the same). I'm assuming that Presence and Awareness are strictly separated, and this duplication is just part of how the example application is built?

Screen Shot 2023-09-20 at 10 57 45@2x

File: packages/liveblocks-core/src/devtools/protocol.ts [hunk 1]
```diff
@@ -91,8 +93,16 @@ export type ClientToPanelMessage =
...
-    };
+    }
 
+  /**
+   * Sent whenever the ydoc is updated
+   */
+  | {
+      msg: "room::sync::ydoc";
+      roomId: string;
+      update: YDocUpdate | UpdateYDocClientMsg;
```
Comment (nvie) [inline packages/liveblocks-core/src/devtools/protocol.ts:104]: Can we rename `YDocUpdate` to `YDocUpdateServerMsg`? I had to look up what it meant, and noticed it's the only one in the union without this suffix:

https://github.com/liveblocks/liveblocks/blob/a1cd3002aa42ff0e29023693b98d38c1fd0e33b7/packages/liveblocks-core/src/protocol/ServerMsg.ts#L29-L51

**Comment by jrowny** [comment]
@nvie Long term there's more work to come on this, but I think most of the problems are with the example more than devtools. 

**Flickering bug:**
Awareness updates are completely controlled by the editor provider, we can check the example, but if it's pointing out a problem in the example, then that's actually DevTools being useful! 

**Infinite loop:**
I think this is another case of DevTools finding a bug, DevTools doesn't communicate anything back, so there would be no way that i know of for DevTools to cause this. We should investigate the example further.

**DocView:**
This needs some longer term work for sure, it makes very little sense in the case of a single large text field, but that's also why we provide other ways to view the data. The reason you don't see "default" in the source code is because TipTap/PM take care of that internally and you only give it the doc. In other editor examples, it's more explicit.

**ChangesList**
 I will work on completely parsing more of the changes so we don't see [object, object], but this is coming directly from Yjs' "toJSON" feature, so it requires custom implementations that go beyond Yjs' various toJSON methods. I think the Content types seen in the list show a good amount of how Yjs internals work just by experimenting with them, I'm not sure how we could make that more friendly, but I'm open to ideas. Internally, Yjs has a "logUpdate' function that's used for debugging that shows basically this same data, so this is just a more friendly way of reading what a developer working with Yjs might already be trying to use.
 
**Awareness View**
The overlap is because the example uses its own data and tiptap also manages the awareness itself. This is a nice advantage of using Liveblocks, as you can customize presence without needing to customize the provider used by TipTap, but it also means there's potentially duplicated data.

File: packages/liveblocks-devtools/src/devtools/panel.html [hunk 2]
```diff
@@ -3,6 +3,7 @@
   <head>
     <meta charset="utf-8" />
     <title>Liveblocks</title>
+    <link href="reactflow/dist/base.css" rel="stylesheet" />
```
[the remark this answers is not in the document]
Comment (jrowny) [reply packages/liveblocks-devtools/src/devtools/panel.html:6]: I think it depends how plasmo is building this, as this comes from a module. I was actually surprised this worked and then I didn't want to touch it :p

File: packages/liveblocks-core/src/devtools/protocol.ts (hunk 1 above)
Comment (jrowny) [reply packages/liveblocks-core/src/devtools/protocol.ts]: fixed, note: will need to update server side at some point.

File: packages/liveblocks-devtools/src/lib/ydoc.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  | string
+  | undefined
+  | unknown[];
+
+export type YType = "unknown" | "text" | "array" | "xml";
+
+/*
+given an item in a ydoc, try and infer its type
+*/
+function getType(item: Y.Item): YType {
```
[the remark this answers is not in the document]
Comment (jrowny) [reply packages/liveblocks-devtools/src/lib/ydoc.ts]: done

**Comment by nvie** [comment]
> **Flickering bug:** Awareness updates are completely controlled by the editor provider, we can check the example, but if it's pointing out a problem in the example, then that's actually DevTools being useful!

In my screen recording, I'm just scrolling up and down inside the panel. Those aren't incoming awareness updates, because nothing happens if my mouse is idle. Only when I start scrolling, the tree view seems to be remounting new rows, causing the highlight to trigger. Probably a bug for @marcbouchenoire to take a look at?

> **Infinite loop:** I think this is another case of DevTools finding a bug, DevTools doesn't communicate anything back, so there would be no way that i know of for DevTools to cause this. We should investigate the example further.

Nice! I've just asked @CTNicholas to take a look at the example. I could actually replicate this on `main`, so this does seem to be an application bug, not a DevTools issue 👍

> **DocView:** This needs some longer term work for sure, it makes very little sense in the case of a single large text field, but that's also why we provide other ways to view the data. The reason you don't see "default" in the source code is because TipTap/PM take care of that internally and you only give it the doc. In other editor examples, it's more explicit.

👍 

> **ChangesList** I will work on completely parsing more of the changes so we don't see [object, object], but this is coming directly from Yjs' "toJSON" feature, so it requires custom implementations that go beyond Yjs' various toJSON methods. I think the Content types seen in the list show a good amount of how Yjs internals work just by experimenting with them, I'm not sure how we could make that more friendly, but I'm open to ideas. Internally, Yjs has a "logUpdate' function that's used for debugging that shows basically this same data, so this is just a more friendly way of reading what a developer working with Yjs might already be trying to use.

👍 

> **Awareness View** The overlap is because the example uses its own data and tiptap also manages the awareness itself. This is a nice advantage of using Liveblocks, as you can customize presence without needing to customize the provider used by TipTap, but it also means there's potentially duplicated data.

👍

**Comment by jrowny** [comment]
@nvie @marcbouchenoire I did some more digging on this one and the root component is not rendering again, presence is not changing, I'm pretty sure it's a CSS transition from the AutoSizer component, but I don't know why it doesn't happen on regular storage view.

**Comment by marcbouchenoire** [comment]
[The animation](https://github.com/liveblocks/liveblocks/blob/e9d68809411c2a5c27fd05dc3ccb84bdb6f6c32d/packages/liveblocks-devtools/src/devtools/components/Tree.tsx#L286-L302) is triggered by deep comparing the row's data, maybe with Yjs the data isn't stable?

**Comment by jrowny** [comment]
> [The animation](https://github.com/liveblocks/liveblocks/blob/e9d68809411c2a5c27fd05dc3ccb84bdb6f6c32d/packages/liveblocks-devtools/src/devtools/components/Tree.tsx#L286-L302) is triggered by deep comparing the row's data so I don't understand how that could be triggered from elsewhere. 🤔

After struggling with this for awhile, I realized that this actually happens in the regular storage tree, so this isn't related to devtools.

**Comment by CTNicholas** [comment]
What do you think about having text wrapping on the pop-out?

[KUNW9yVBOX.webm](https://github.com/liveblocks/liveblocks/assets/33033422/7279dc2d-3eb7-4d13-ad79-8f0b9006830b)

**Review by nvie — APPROVED** [review]
Looking good from the protocol side, and also from the UX side! Great team work, @jrowny + @marcbouchenoire 👌 !

**Review by marcbouchenoire — APPROVED** [review]
LGTM!

**Comment by nvie** [comment]
@jrowny I'm trying to implement the room event visualizer feature ask from Chris today, since both room event updates and DevTools updates are ready to released. It would be nice to wrap this all up in a single release. Maybe first release this as 1.4 and then push the DevTools release out?

**Pull request merged**